report/report.py

```diff
@@ -0,0 +1,178 @@
+from sisyphus import *
+
+import sys
+import getpass
+import gzip
+from datetime import datetime
+from typing import Dict, Union, Callable
+
+_Report_Type = Dict[str, Union[tk.AbstractPath, str]]
+
+
+class ReportResultsJob(Job):
+    """
+    Job to report results either via mail and/or just condensed in a file
+    """
+
+    def __init__(
+        self,
+        name: str,
+        report: _Report_Type,
+        report_format: Callable[[_Report_Type, any], str],
+        compress: bool = True,
+        **report_format_kwargs,
+    ):
+        """
+
+        :param name: Name of the report
+        :param report: Dictionary containing the report files
+        :param report_format: Function that converts the report dictionary to a string
+        :param report_format_kwargs: Addtional kwargs for the report format function
+        :param compress: Whether to compress the report or not
+        """
+        self.name = name
+        self.report = report
+        self.report_format = report_format
+        self.report_format_kwargs = report_format_kwargs
+        self.compress = compress
+
+        self.sis_command_line = sys.argv
+        self.mail_address = getattr(gs, "MAIL_ADDRESS", None)
+        self.config_path = tk.config_manager.current_config
+        assert self.config_path is not None, "Could not find config path"
+
+        for i in ["date", "user", "name", "config_path", "sis_command_line"]:
+            assert i not in report, "%s will be set automatically"
+
+        self.out_report = self.output_path("report.gz")
+
+    def run(self):
+        user = getpass.getuser()
+        report = self.report.copy()
+
+        report["date"] = datetime.now().strftime("%d/%m/%Y %H:%M:%S")
+        report["user"] = user
+        report["name"] = self.name
+        report["config_path"] = self.config_path
+        report["sis_command_line"] = str(self.sis_command_line)
+
+        if self.compress:
+            with gzip.open(self.out_report.get_path(), "w") as f:
+                f.write(
+                    self.report_format(report, **self.report_format_kwargs).encode()
+                    + b"\n"
+                )
+        else:
+            with open(self.out_report.get_path(), "w") as f:
+                f.write(self.report_format(report, **self.report_format_kwargs) + "\n")
```

Comment:
Is it intentional that one branch encodes the string and the other doesn't? Is it required by `gzip.open` vs `open`?
Or is it a problem of specifying mode "wt" vs "wb" explicitly?

Reply:
Tbh. I dont know. The binary variant is copied from the old code and the non compressed version was added by me. But from my understanding (which might be off by a longshot, I dont have much experience with this) I think "wb" expects some kind of binary input and does not convert "internally".

Reply:
see https://docs.python.org/3/library/functions.html#open
> [mode] defaults to 'r' which means open for reading in text mode.

and https://docs.python.org/3/library/gzip.html
> The mode argument can be any of 'r', 'rb', 'a', 'ab', 'w', 'wb', 'x' or 'xb' for binary mode, or 'rt', 'at', 'wt', or 'xt' for text mode. The default is 'rb'.

so 'w' for builtin `open` defaults to text mode and 'w' for `gzip.open` defaults to binary. I suggest to decide which one you want and then use 'wb' or 'wt' consistently.